### PR TITLE
Support for iron opcodes

### PR DESCRIPTION
**Pull request by vyzo**
Companion to https://github.com/filecoin-project/builtin-actors/pull/713

TODO:
- [x] `syscalls::gas::available`
- [x] `syscalls::actor::actor_balance` -- renamed to `syscalls::actor::balance_of`
- [x] `syscalls::env::timestamp`
- [x] `syscalls::env::blockhash`
- [x] ~`syscalls::env::gas_limit`~ -- moved into the invocation context.
- [x] ~`syscalls::env::gas_price`~ -- added gas_premium to the invocation context.

### CODE REVIEW

File: fvm/src/syscalls/actor.rs
Comment (raulk) [inline fvm/src/syscalls/actor.rs:123]: `balance_of` sounds like a better name to prevent the stutter (`actor::actor_balance`).

This seems relatively easy to implement.

File: fvm/src/syscalls/env.rs
Comment (raulk) [inline fvm/src/syscalls/env.rs:6]: This will require the chain environment data to be passed from the client.

File: fvm/src/syscalls/env.rs
Comment (raulk) [inline fvm/src/syscalls/env.rs:8]: This should be `tipset_cid`, the block hash is something different than what this should return.

File: fvm/src/syscalls/env.rs
Comment (raulk) [inline fvm/src/syscalls/env.rs:4]: `tipset_timestamp`?

Comment (raulk) [inline fvm/src/syscalls/env.rs:4]: Just noting that we had discussed to put all this data behind a single struct, following the same patter as the `InvocationContext`, e.g. `ChainContext`. I'm happy to keep it as separate ops for now, it's always easy to add another syscall to retrieve all info at once for when the actor wants more than just a single field.

File: fvm/src/syscalls/env.rs
Comment (raulk) [inline fvm/src/syscalls/env.rs:1]: Not sure if `env` is the appropriate module name here. Could also be `chain`.

File: fvm/src/syscalls/env.rs
Comment (raulk) [inline fvm/src/syscalls/env.rs:23]: This is not correct. We already have the base_fee under `network::base_fee`. The EVM `GASPRICE` opcode returns the `effective_gas_price`, which is:

	effective_gas_price = priority_fee_per_gas + block.base_fee_per_gas

The piece we're missing is the gas premium (in the invocation context?), and the EVM runtime would fetch both the base fee and the gas premium to sum them and honour the GASPRICE behaviour.

File: fvm/src/syscalls/gas.rs
Comment (raulk) [inline fvm/src/syscalls/gas.rs:22]: Should be trivial to implement. `context.kernel.gas_available()` already exists.

File: fvm/src/syscalls/actor.rs
Comment (Stebalien) [inline fvm/src/syscalls/actor.rs:120]: I'd consider:

```suggestion
    actor_id: u64,
```

This will require an extra syscall to lookup the actor ID in some cases, but it matches our other syscalls. All the calls that take an _address_ do so because the target actor may not exist (yet).

File: fvm/src/syscalls/env.rs
Comment (Stebalien) [inline fvm/src/syscalls/env.rs:10]: `epoch: i64`

File: fvm/src/syscalls/env.rs
Comment (Stebalien) [reply fvm/src/syscalls/env.rs:23]: (also needs to be a token amount)

File: fvm/src/syscalls/env.rs
Comment (Stebalien) [reply fvm/src/syscalls/env.rs:4]: IMO, we'd also want to move some fields out of the invocation context (epoch, network version, etc.) into the ChainContext (or EpochContext?).

(I'd prefer todo this _sooner_ rather than later, if possible)

**Comment by mriise** [comment]
There are overlaps here with what I need to do for ValidateContext (if at all). Some fields in `InvokationContext` are more related to the message or the network, I think at this point we can just make a `xContext` struct per module to keep seperation. This would split current `InvokationContext` into `MessageContext` and `NetworkContext`, the latter might just be combined with the this PR's `ChainContext` depending on what `Ops` it needs. For syscalls that are added here as one offs, I think we should leave them as-is.

Doing it like this sort of deflates the purpose of context, but having consolidated structs of metadata is useful still, and we can continue to cache the result in the sdk as needed.

File: fvm/src/syscalls/actor.rs
Comment (vyzo) [reply fvm/src/syscalls/actor.rs:123]: ok, will rename.

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs:8]: sure!

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs:4]: ok.

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs:1]: we can rename, let's discuss. I am keeping it for now, but let's resolve before merging.

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs:23]: so you want to remove this syscall entirely? I can add the gas premium in the invocation context.

File: fvm/src/syscalls/actor.rs
Comment (vyzo) [reply fvm/src/syscalls/actor.rs:120]: ok, will use the actor id.

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs:10]: ACK

File: fvm/src/syscalls/actor.rs
Comment (vyzo) [reply fvm/src/syscalls/actor.rs]: done and implemnted.

File: fvm/src/syscalls/actor.rs
Comment (vyzo) [reply fvm/src/syscalls/actor.rs]: done.

File: fvm/src/syscalls/gas.rs
Comment (vyzo) [reply fvm/src/syscalls/gas.rs]: implemented.

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs:23]: I can also move the gas limit to the invocation context as well, since they are intrinsically related.

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs:4]: Sure, but I'd rather do this in a subsequent pr.

File: fvm/src/syscalls/env.rs
Comment (vyzo) [reply fvm/src/syscalls/env.rs]: Ok, instead of passing them in the env context, I added gas premium as an argument to Send and fished the limit from the gas tracker.
I also added both of them to the invocation context struct.

File: fvm/src/call_manager/default.rs
Comment (raulk) [inline fvm/src/call_manager/default.rs:109]: This looks awkward and leaky here. All parameters to send are **necessary** to execute the send; this is merely contextual information. Let's not put it here, but instead extend the `GasTracker` to hold this value and add a getter for it.

File: fvm/src/call_manager/default.rs
Comment (raulk) [inline fvm/src/call_manager/default.rs:417]: Doesn't the Kernel already have access to the `gas_limit` via the `call_manager.gas_tracker`? Why does it need to be handed off the field explicitly? (for premium, see above)

File: fvm/src/call_manager/mod.rs
Comment (raulk) [inline fvm/src/call_manager/mod.rs:51]: The ChainContext is a Machine level field, not a CallManager level field. Let's hoist it up and remove it from here. (The ChainContext won't change across message applications within a single Machine instantiation)

Comment (raulk) [reply fvm/src/call_manager/mod.rs:51]: We can also move the `epoch` into the chain context, like @Stebalien suggested elsewhere.

File: fvm/src/kernel/default.rs
Comment (raulk) [inline fvm/src/kernel/default.rs:476]: For extra safety and formality, we should _also_ validate that the specific lookback is within the admitted range (256 for Ethereum, 900 if we want to follow Filecoin's finality).

File: fvm/src/kernel/mod.rs
Comment (raulk) [inline fvm/src/kernel/mod.rs:85]: See above. Can get the limit from the `call_manager.gas_tracker`, and we should add the premium there too.

File: fvm/src/kernel/mod.rs
Comment (raulk) [inline fvm/src/kernel/mod.rs:364]: For better naming consistency.

```suggestion
pub trait ChainOps {
```

File: sdk/src/sys/env.rs
Comment (raulk) [inline sdk/src/sys/env.rs:5]:
```suggestion
    module = "chain";
```

**Review by raulk — CHANGES_REQUESTED** [review]

File: fvm/src/call_manager/default.rs
Comment (vyzo) [reply fvm/src/call_manager/default.rs:417]: oh, didn't notice that -- yeah no need for it then.

File: fvm/src/call_manager/default.rs
Comment (vyzo) [reply fvm/src/call_manager/default.rs:109]: ok.

File: fvm/src/call_manager/mod.rs
Comment (vyzo) [reply fvm/src/call_manager/mod.rs:51]: Ok, will hoist.

File: fvm/src/kernel/default.rs
Comment (vyzo) [reply fvm/src/kernel/default.rs:476]: this should probably happen at call site? We don't know if it is an eth or fil invocation here, so we merely check against the length of the vector.

File: fvm/src/kernel/mod.rs
Comment (vyzo) [reply fvm/src/kernel/mod.rs:85]: ack.

File: fvm/src/kernel/mod.rs
Comment (vyzo) [reply fvm/src/kernel/mod.rs:364]: ok.

File: fvm/src/kernel/default.rs
Comment (raulk) [reply fvm/src/kernel/default.rs:476]: The FVM should place a hard limit. I would vote for 900 epochs because it matches our finality. The EVM runtime should validate that the argument is 256 max.

**Comment by raulk** [comment]
BTW @vyzo -- I agree with @mriise's comment above. Thinking about it, all attributes in `ChainContext` are actually network-related attributes that _must_ be provided by the client, so we can consolidate into the `network` syscall module and package, and into a `NetworkContext`, instead of accepting the package/namespace sprawl here.

**Comment by vyzo** [comment]
ok

File: fvm/src/call_manager/default.rs
Comment (vyzo) [reply fvm/src/call_manager/default.rs:109]: done.

File: fvm/src/kernel/mod.rs
Comment (vyzo) [reply fvm/src/kernel/mod.rs]: done.

File: fvm/src/call_manager/default.rs
Comment (vyzo) [reply fvm/src/call_manager/default.rs]: done.

**Comment by vyzo** [comment]
hoisted chain context stuffs into a NetworkContext struct in Machine, which includes the epoch as well.

File: fvm/src/call_manager/mod.rs
Comment (vyzo) [reply fvm/src/call_manager/mod.rs]: done.

File: fvm/src/kernel/mod.rs
Comment (vyzo) [reply fvm/src/kernel/mod.rs:361]: done.

File: sdk/src/sys/env.rs
Comment (vyzo) [reply sdk/src/sys/env.rs:5]: done.

File: fvm/src/kernel/default.rs
Comment (vyzo) [reply fvm/src/kernel/default.rs]: ok, 900 it is -- so we get an error (as opposed to a None) if we pass more than that.

**Review by raulk — COMMENTED** [review]
_Almost_ right, just a bit more consolidation (NetworkOps and moving base_fee and nv into the NetworkContext), and we're good.

File: fvm/src/call_manager/default.rs
Comment (vyzo) [reply fvm/src/call_manager/default.rs:79]: yeah, make sense.

File: fvm/src/kernel/default.rs
Comment (vyzo) [reply fvm/src/kernel/default.rs:477]: yeah, i guess i should!

File: fvm/src/machine/mod.rs
Comment (vyzo) [reply fvm/src/machine/mod.rs:208]: sure, why not.

Comment (vyzo) [reply fvm/src/machine/mod.rs:208]: I moved the base fee, but the nv is rather awkward to move -- it is in the NetworkConfig struct and I think it might be best to leave it there for now.

File: fvm/src/kernel/mod.rs
Comment (vyzo) [reply fvm/src/kernel/mod.rs:367]: done.

File: fvm/src/machine/mod.rs
Comment (raulk) [inline fvm/src/machine/mod.rs:195]: There's more refactor/consolidation we'll need to do here between `NetworkConfig` and `NetworkContext`.

`NetworkConfig` is confusing because it actually contains **machine** configuration:

- max call depth
- max wasm stack
- actor debugging
- actor redirects
- builtin overrides

As well as some network attributes which should live in `NetworkContext`:

- network version
- price list

But we don't need to do it _now_.

**Review by raulk — APPROVED** [review]

**Pull request merged**
